apps/sms/services/js/message_manager.js

```diff
@@ -429,26 +429,23 @@ var MessageManager = {
     // other potential call to the API, like the one for getting a message
     // list, could be done within the calls to mark the messages as read.
 
-    var id = list.pop();
-    // TODO: Third parameter of markMessageRead is return read request.
-    //       Here we always return read request for now, but we can let user
-    //       decide to return request or not in Bug 971658.
-    var req = this._mozMobileMessage.markMessageRead(id, isRead, true);
-    // isRead == false i.e mark as unread case, marking only one message
-    // as unread is sufficient.
-    req.onsuccess = (function onsuccess() {
-      if (!list.length || !isRead) {
-        return;
-      } else if (isRead) {
-        this.markMessagesRead(list, isRead);
+    // shouldSendReadReport == true if Send-Read-Reports switch in
+    //  Messaging Settings is enabled, else false
+
+    var markMessage = (shouldSendReadReport) => {
+      if (!list.length) {
+        return Promise.resolve();
       }
-    }).bind(this);
 
-    req.onerror = function onerror() {
-      console.error(
-        'Error while marking message %d as read: %s', id, this.error.name
+      var id = list.pop();
+      var req = this._mozMobileMessage.markMessageRead(id, isRead, shouldSendReadReport);
+
+      return req.then(
+        () => isRead ? markMessage(shouldSendReadReport) : Promise.resolve()
       );
     };
+
+    return Settings.sendReadReport.then(markMessage);
```

Comment:
If we treat the sendReadReport as a simple boolean, we can just return markMessage() and no need to provide any parameter for markMessage. You can return
```
return this._mozMobileMessage.markMessageRead(
  id,
  isRead,
  Settings.shouldSendReadReport
).then(
  ...
```